core/src/main/java/org/mule/runtime/core/internal/policy/PolicyNextActionMessageProcessor.java

```diff
@@ -58,113 +48,73 @@
-        .flatMap(event -> subscriberContext()
-            .flatMap(ctx -> just(ctx.hasKey(POLICY_IS_PROPAGATE_MESSAGE_TRANSFORMATIONS)
-                ? policyEventConverter.createEvent((PrivilegedEvent) event,
-                                                   getOriginalEvent(event),
-                                                   builder -> saveState(builder, event),
-                                                   (boolean) ctx.get(POLICY_IS_PROPAGATE_MESSAGE_TRANSFORMATIONS))
-                : policyEventConverter.createEvent((PrivilegedEvent) event,
-                                                   getOriginalEvent(event),
-                                                   builder -> saveState(builder, event)))))
+                                                   coreEvent.getMessage(), getPolicyId()))
         .doOnNext(event -> {
           popBeforeNextFlowFlowStackElement().accept(event);
           notificationHelper.notification(BEFORE_NEXT).accept(event);
         })
         .compose(eventPub -> subscriberContext()
-            .flatMapMany(ctx -> eventPub.transform(ctx.get(POLICY_NEXT_OPERATION)).cast(CoreEvent.class)))
+            .flatMapMany(ctx -> eventPub
+                .flatMap(event -> ctx.hasKey(POLICY_IS_PROPAGATE_MESSAGE_TRANSFORMATIONS)
+                    ? just(policyEventMapper.onSourcePolicyNext(event, ctx.get(POLICY_IS_PROPAGATE_MESSAGE_TRANSFORMATIONS)))
+                    : just(policyEventMapper.onOperationPolicyNext(event)))
+                .transform(ctx.get(POLICY_NEXT_OPERATION))
+                .cast(CoreEvent.class)))
         .doOnNext(coreEvent -> {
           notificationHelper.fireNotification(coreEvent, null, AFTER_NEXT);
           pushAfterNextFlowStackElement().accept(coreEvent);
           logExecuteNextEvent("After execute-next", coreEvent.getContext(), coreEvent.getMessage(),
-                              this.muleContext.getConfiguration().getId());
+                              getPolicyId());
         })
-        .map(result -> (CoreEvent) policyEventConverter.createEvent((PrivilegedEvent) result,
-                                                                    loadState((PrivilegedEvent) result)))
-        .onErrorContinue(MessagingException.class, (error, ev) -> {
-          final CoreEvent event = ((MessagingException) error).getEvent();
-
-          if (isEventContextHandledByThisNext(event)) {
-            MessagingException me = (MessagingException) error;
-            notificationHelper.fireNotification(me.getEvent(), me, AFTER_NEXT);
-            pushAfterNextFlowStackElement().accept(me.getEvent());
-
-            me.setProcessedEvent(policyEventConverter.createEvent((PrivilegedEvent) me.getEvent(),
-                                                                  loadState((PrivilegedEvent) me.getEvent())));
+        .map(result -> policyEventMapper.fromPolicyNext(result))
+        .onErrorContinue(MessagingException.class, (error, v) -> {
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove useless curly braces around statement ([squid:S1602](https://sonarqube.kbuild.msap.io/coding_rules#rule_key=squid%3AS1602))

[See it in SonarQube](https://sonarqube.kbuild.msap.io/project/issues?id=mule-runtime-mule&pullRequest=7763&issues=AWnLHoR7otIXnmsUntKA&open=AWnLHoR7otIXnmsUntKA)

[ISSUE_KEY]: AWnLHoR7otIXnmsUntKA